idseq_dag/util/log.py

```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
         if flush:
             sys.stdout.flush()
 
+def log_event(event_name, values=None, start_time=None, warning=False, flush=True):
+    '''
+    Write a new log event.
+
+    Parameters:
+    event_name (str): name of the event
+    values(dict): Optional. Values associated to that event. Will be logged in json format.
+    start_time(datetime): Optional. If given, it will calc the elapsed time from this datetime to now and write it to the log line.
+
+    Returns:
+    datetime: Now. It can be used to pass to parameter start_time in a future log_event call.
+
+    Example:
+    start = log_event("downloaded_started", {"file":"abc"})
+    # ... download your file here ...
+    log_event("downloaded_completed", {"file":"abc"}, start_time=start)
+    '''
+    fmt_values = "{}" if values is None else json.dumps(values)
+    if start_time is None:
+        fmt_message = "Event '%s' %s" % (event_name, fmt_values)
+    else:
+        duration = (datetime.datetime.now()-start_time).total_seconds()
+        fmt_message = "Event '%s' %s (%.1f seconds)" % (event_name, fmt_values, duration)
+    write(fmt_message, warning, flush)
+    return datetime.datetime.now()
+
+@contextmanager
+def log_context(context_name, values=None, log_caller_info=True):
+    '''
+    Log context manager to track started and completed events for a block of code
+
+    Parameters:
+    context_name (str): Name for this context.
+    values(dict): Optional. Values associated to that event. Will be logged in json format.
+    log_caller_info(bool): Log the caller file_name and method name.
+
+    Example:
+    # some_file.py
+    from log import log_context
+
+    def some_method
+        # ... some code here ...
+        with log_context("sub_step", {"abc": "123"}):
+            # ... your code goes here ...
+
+    The code above will generate two log entries:
+    Event ctx_start {"n": "sub_step", "v": {abc": 123}, "caller": {"f": ["some_file.py", "some_method"]}
+    Event ctx_end {"n": "sub_step", "v": {abc": 123}, "caller": {"f": ["some_file.py", "some_method"]} (0.3 sec)
+    '''
+    val = {"n": context_name}
+    if values is not None:
+        val["v"] = values
+    if log_caller_info:
+        f_code = sys._getframe(2).f_code
+        val["caller"] = {"f": [os.path.basename(f_code.co_filename), f_code.co_name]}
+    start = log_event("ctx_start", val)
+    try:
+        yield
+        log_event("ctx_end", val, start_time=start)
+    except Exception as e:
+        val["error_type"] = type(e).__name__
+        val["error_args"] = e.args
+        log_event("ctx_error", val, start_time=start)
+        raise e
```

Comment:
This is one of those occasions where I do not agree with re-raising. 
Logging should be itself a parallel and non-intrusive action to the main logic.

Reply:
This re-raise is exactly there to be non-intrusive. This catch block is catching an error on the wrapped method and re-raising the same error.

Reply:
I am just saying that it is not responsability of a log util to do that.
Here you should catch and handle errors on the log function and not the others.
Also, I believe this will add the log to the stack trace, adding to the confusion. Can you confirm that?

Reply:
@tfrcarvalho :
It keeps the same info:
![image](https://user-images.githubusercontent.com/8084076/56986302-3fdbec80-6b3f-11e9-8153-1ab5ca5770c6.png)
```python
import traceback

from contextlib import contextmanager

@contextmanager
def log_context():
    print("ctx_start")
    try:
        yield
        print("ctx_end")
    except Exception as e:
        print("ctx_error")
        raise e

def throws_error():
  1 / 0 # Division by zero

def test1():
  try:
    throws_error()
  except:
    traceback.print_exc()

def test2():
  try:
    with log_context():
      throws_error()
  except:
    traceback.print_exc()

print("-------- test1() stacktrace --------")
test1()
print();
print("-------- test2() stacktrace --------")
test2()
```

Reply:
The idea here is not hiding errors or even use this feature to replace the original error handlers. You will notice that I'm not even printing the stack trace, because I feel this is a task for the caller.
The idea of this log context is to provide a way to know when a code block started and ended, and how long it took. The problem is if we don't catch exceptions here (and re-raise them), we will never know if the code has failed or if it has never completed.

Reply:
The new  PR `log_context` provides a way to know when a code block started and ended, and how long it takes. The problem is if we don't catch exceptions here (and re-raise them), we will never know if the code has failed or if it has never completed.
The idea is not hiding errors or even use this feature to replace the original error handlers. You will notice that I'm not even printing the stack trace, because I feel this is a task for the caller.

Reply:
I think it is fine then.